### ISSUE TITLE
Road to 1.0 Redux

### ISSUE THREAD

**Issue by basarat** [typestyle/typestyle#59]
So the core is stable enough in my mind to go to `1.0`. This was tracked by https://github.com/typestyle/typestyle/issues/43 and is essentially done. 

The plan for people using TypeStyle for creating *libraries* after 1.0 is to have them `peerDependency : typestyle@1.*.*`. And then the person using the library gets to chose the latest/greatest TypeStyle 1.x.

However we also ship `typestyle/lib/csx` and there are concerns about what the API for that should be e.g. recently raised https://github.com/typestyle/typestyle/issues/53 and we don't want to release anything that can be a breaking change too often after `1.0`. Due to the nature of what `csx` currently contains this can be a problem so this issue is going to track us cleaning up `csx`

* [x] Page setup stays. 
* [x] `colors` stay :rose: 
* [x] Functions named after any `CSSFunction` (e.g. `rgb`) stay. 
* [x] Everything else needs a new home moved to csstips : https://github.com/typestyle/csstips. e.g. 
    *  `verticallySpaced`, *layer based layouts* 
    * Functions that return objects that can be *mixed in* named after CSSProperties e.g. `margin` / `padding` 
    * flex box , Happy to rename current flexbox stuff to match other functions that return CSSProperties that can be mixed in e.g. `flex` should be a function that creates an object with the `flex` property (currently it is more like a library). This can be 🚲 🏠 (bike shedded) after we move it out.

/cc @notoriousb1t @jamiewinder :rose:

**Comment by basarat** [comment]
Notes to self from our discussions on gitter: 

* Consider `px` and `percent` functions instead of mixins for `width`/`height` etc. Example : `width: csx.px(10)`
* Consider constants e.g. `fontWeight: font.normal` 

_ I will keep this updated for my reference_ :rose:

**Comment by jamiewinder** [comment]
> Happy to rename current flexbox stuff to match other functions that return CSSProperties that can be mixed in e.g. flex should be a function that creates an object with the flex property (currently it is more like a library).

Something like the following helpers:
```typescript
const flex = (value: string) => ({
   '-webkit-flex': value,
   '-ms-flex': value,
   flex: value,
});

const myStyle = style({
    color: 'red',
    ...flex('0 0 auto')
});
```
(Perhaps with different parameterisation). Might be handy, especially with the arrival of object spread in TypeScript 2.1 which I personally think makes this more clear than passing multiple objects. The problem will always be the argument over how far the vendor prefixing should go in terms of browser support.

**Comment by basarat** [comment]
> Might be handy, especially with the arrival of object spread in TypeScript 2.1

`...` is a shallow extend. The varargs in style merge multiple nested objects (e.g. hover states).

> The problem will always be the argument over how far the vendor prefixing should go in terms of browser support

Agreed. Flexbox and other mixin creating objects will be moved out into `csstips` for that reason so that people don't need to change `typestyle` versions to get the *right* vendor prefixes *for them*. 

 I've updated the original message with the plan : https://github.com/typestyle/typestyle/issues/59#issue-194237804 :rose:

**Comment by basarat** [comment]
Moved out all mixin / mixin creators into csstips : https://github.com/typestyle/typestyle/commit/93fb81e4c4792a70161961fa394c4164dd65b1b2 
Also moved out opinionated page setup as well : https://github.com/typestyle/typestyle/commit/2c13e9e28ec02c76ba6bfac1fc9a2903981119aa

So now 

* `typestyle` is just that ... *types* + `style` :) 
* And `csx` is just CSSFunctions and CSSValues and named to match just that (e.g. `rgb`,`color` etc)
* Everything else is csstips : https://github.com/typestyle/csstips

I will cut a release in due time + add docs :rose:

**Comment by basarat** [comment]
Released TypeStyle `0.15.0`. Everything other than colors is moved out to csstips : https://github.com/typestyle/csstips 

# Protip for refactoring: 

* Install latest typestyle and csstips. 
* you should get a compile error if you use e.g. `csx.verticallySpaced`. 
* Just go the top of the file and rename `import * as csx from "typestyle/lib/csx";` to `import * as csx from "csstips"`. Now you can use the TypeScript *refactor* to rename `csx` to `csstips`. Easy :rose:

Resetting the timer for release to 1.0 by 1 week. Lets see if we feel more confident about no more breaking changes after that :heart:
